txnkv/transaction/txn_file.go

```diff
@@ -0,0 +1,936 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transaction
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/errorpb"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pkg/errors"
+	"github.com/tikv/client-go/v2/config"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/internal/client"
+	"github.com/tikv/client-go/v2/internal/locate"
+	"github.com/tikv/client-go/v2/internal/logutil"
+	"github.com/tikv/client-go/v2/internal/retry"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+	"github.com/tikv/client-go/v2/tikvrpc"
+	"github.com/tikv/client-go/v2/txnkv/txnlock"
+	"github.com/tikv/client-go/v2/util"
+	atomicutil "go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+var (
+	// BuildTxnFileMaxBackoff is max sleep time (in millisecond) to build TxnFile.
+	BuildTxnFileMaxBackoff = atomicutil.NewUint64(60000)
+
+	buildChunkErrMsg string = "txn file: build chunk failed"
+)
+
+const PreSplitRegionChunks = 4
+
+type txnFileCtx struct {
+	slice txnChunkSlice
+}
+
+type chunkBatch struct {
+	txnChunkSlice
+	region    *locate.KeyLocation
+	isPrimary bool
+}
+
+func (b chunkBatch) String() string {
+	return fmt.Sprintf("chunkBatch{region: %s, isPrimary: %t, txnChunkSlice: %s}", b.region.String(), b.isPrimary, b.txnChunkSlice.String())
+}
+
+// chunkBatch.txnChunkSlice should be sorted by smallest.
+func (b *chunkBatch) getSampleKeys() [][]byte {
```

Comment:
How about rename to `getSortedChunkSliceSmallestKeys`, which better represents what the function returns.

Reply:
The method is named as its purpose, not implementation.

And the implementation would be changed in the future, e.g, we can store extra keys in the middle of chunk and use here, if necessary.